src/prove.rs

```diff
@@ -1,88 +1,223 @@
 use aes::{
     cipher::{BlockEncrypt, KeyInit},
-    Aes128,
+    Aes128, Aes256,
 };
-use cipher::block_padding::NoPadding;
-use std::sync::mpsc;
+use cipher::{block_padding::NoPadding, generic_array::GenericArray};
+use eyre::Context;
+
+use std::{collections::HashMap, ops::Range, path::Path};
+use streaming_iterator::StreamingIterator;
+
+use crate::{config::Config, difficulty::proving_difficulty, metadata, reader::stream_data};
 
 const BLOCK_SIZE: usize = 16; // size of the aes block
-const BATCH: usize = 8; // will use encrypt8 asm method
-const CHUNK_SIZE: usize = BLOCK_SIZE * BATCH;
+const AES_BATCH: usize = 8; // will use encrypt8 asm method
+const CHUNK_SIZE: usize = BLOCK_SIZE * AES_BATCH;
 
-pub struct Prover<const N: usize = 1> {
-    ciphers: [Aes128; N],
-    output: [u8; CHUNK_SIZE],
-    d: u64,
+#[derive(Debug)]
+pub struct Proof {
+    pub nonce: u32,
+    pub indicies: Vec<u64>,
 }
 
-impl<const N: usize> Prover<N> {
-    pub fn new(challenge: &[u8; 16], d: u64) -> Self {
-        let ciphers: [Aes128; N] = (0..N as u32)
-            .map(|i: u32| {
-                let mut key = [0u8; BLOCK_SIZE];
-                key[..12].copy_from_slice(&challenge[..12]);
-                key[12..].copy_from_slice(&i.to_le_bytes());
-                Aes128::new(&key.into())
-            })
-            .collect::<Vec<_>>()
-            .try_into()
-            .unwrap();
-        let output = [0u8; CHUNK_SIZE];
-        Prover { ciphers, output, d }
+pub trait Prover<'a> {
+    fn prove<F>(&self, batch: &[u8], index: u64, consume: F) -> Option<u32>
+    where
+        F: FnMut(u32, u64) -> bool;
+
+    fn required_aeses(&self) -> usize;
+}
+
+#[derive(Debug)]
+struct AesCipher {
+    pub aes: Aes128,
+    pub nonce: u32,
+}
+
+impl AesCipher {
+    fn new(challenge: &[u8; 32], nonce: u32) -> Self {
+        let key_cipher = Aes256::new(challenge.into());
```

Comment:
why do you need AES256? i think this is some kind of misunderstanding with research team

Reply:
I use AES256 just to encrypt the nonce with the challenge because AES256 uses a 32B key. It's the same in Go implementation.

Reply:
i understood it, but this i am not sure that it suppose to be done this way. i would expect that stuff to be simply hashed with blake3 (since we are already using it)

Reply:
This must be sorted out with the research